keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+
+
+### Add .status.lastSuccessfulTime
+[#issue/75674](https://github.com/kubernetes/kubernetes/issues/75674)  
+Add `lastSuccessfulTime` to `.status` that tracks the last time the job completed successfully. This will augment the `lastScheduledTime` available in the `.status` in the v1beta1 api. Potential use is in monitoring (e.g. fire an alert if lastSuccessfulTime is more than X ago).
+
+Code for reference, subject to change:
+
+```golang
+func (cjc *CronJobController)  syncOne(sj *batchv1beta1.CronJob, js []batchv1.Job, now time.Time, jc jobControlInterface, sjc sjControlInterface, recorder record.EventRecorder) {
+   ...
+   // Get most recent successfully finished job from current job list.
+   sj.Status.LastSuccessfulTime = &metav1.Time{Time: scheduledTime}
+	if _, err := sjc.UpdateStatus(sj); err != nil {
+		klog.Infof("Unable to update status for %s (rv = %s): %v", nameForLog, sj.ResourceVersion, err)
+	}
+   ...
+}
+```
+
+### Add .status.nextScheduleTime
+[#issue/78564](https://github.com/kubernetes/kubernetes/issues/78564)
+Sdd `nextScheduleTime` to `.status` that tracks the next time the job will be scheduled. This may not be accurate with `Forbid` concurrency policy. This only tracks the `Job` creation time and not the actual `Pod` creation time.
+
+| Concurrency policy | notes |
+|:----------|:----------|
+| Allow   | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter |
+| Forbid  |  `nextScheduleTime` would not be accurate if the previous `Job` takes longer than the cron interval time |
+| Replace | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter along with older concurrent `Job` cleanup if applicable. |
+
+
+```golang
+func (cjc *CronJobController)  syncOne(sj *batchv1beta1.CronJob, js []batchv1.Job, now time.Time, jc jobControlInterface, sjc sjControlInterface, recorder record.EventRecorder) {
+   ...
+   // Find next schedule time
+   sj.Status.NextScheduleTime = &metav1.Time{Time: scheduledTime}
+	if _, err := sjc.UpdateStatus(sj); err != nil {
+		klog.Infof("Unable to update status for %s (rv = %s): %v", nameForLog, sj.ResourceVersion, err)
+	}
+   ...
+}
+```
+
+### Add Counters
+Add a set of counters that helps users understand a summary of runs.  
+- `SuccessfulRuns` Count of all successful runs
+- `FailedlRuns` Count of all failed runs
+- `FailuresSinceSuccess` Count of all successful runs since last failure
+
+### Add .status.conditions
+Add a condition array with `Settled` condition type. This would help with the effort of standarizng conditions across all core types. `Settled` is set at the end of every successful reconcile run. The key thing to note here is the notions of `Settled` does not imply the `Job`s are running correctly. It just means that the controller is done processing this object successfully.
+
+### CatchUp concurrency support
```

Comment:
Drop it, the fact there's no consensus will make this proposal take even longer. Let's focus on the simple goals we can achieve with this proposal, there are quite a lot of them, already.

Reply:
ack